### PR TITLE
feat: add link operator

### LINKED ISSUE #760: Add a `@link` operator
### Overview
Introducing the `@link` operator for linking and importing of external files in configurations.

#### Proposed GraphQL Schema Definition
Create a link operator that has `src`, `type` and `id` fields as follows.
```graphql
schema 
  @link(type: "tailcall/graphql", src: "https://abc.com/graphql")
  @link(id: "abc" type: "tailcall/yml", src: "https://abc.com/application.yml")
  @link(type: "tailcall/graphql", src: "https://abc.com/application.graphql") {
  query: Query
}
```

**Technical Requirements**
- Each file is resolved and inside of config_reader.
- Resolved files are either merged with the original config or attached as an extension in the config_set.

### PR DESCRIPTION
**Pull request by ologbonowiwi**
**Summary:**  
Implements `@link` operator

**Issue Reference(s):**  
Fixes #760
/claim #760

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/config/link.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    pub src: String,
+    ///
+    /// The id of the link. It is used to reference the link in the schema.
+    ///
+    #[serde(default, skip_serializing_if = "is_default")]
+    pub id: Option<String>,
+    ///
+    /// Raw content of the file
+    ///
+    content: Option<String>,
```
Comment (tusharmath) [inline src/config/link.rs:35]: Content isn't required anymore.

Comment (ologbonowiwi) [inline src/config/link.rs:35]:
```suggestion
```

File: src/config/reader.rs [hunk 2]
```diff
@@ -62,6 +65,77 @@ impl ConfigReader {
...
+                    }
+
+                    config_set.extensions.file_descriptor_from_links.push(
+                        FileDescriptorSetWithId {
+                            id: config_link.id.to_owned(),
+                            file_descriptor_set,
+                        },
+                    );
+                }
+                LinkType::Data => {
```
Comment (tusharmath) [inline src/config/reader.rs:127]: Let's not support Data type already. The future of this type isn't clear.

File: tests/graphql/test-add-link-to-empty-config.graphql [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#> server-sdl
+schema @server @upstream @link(src: "tests/graphql/fixtures/link-const.graphql", type: Config) @link(src: "tests/graphql/fixtures/link-enum.graphql", type: Config) {
```
Comment (tusharmath) [inline tests/graphql/test-add-link-to-empty-config.graphql:2]: paths should be relative to current file.

File: src/config/reader.rs [hunk 4]
```diff
@@ -62,6 +65,77 @@ impl ConfigReader {
...
+                    config_set = config_set.merge_right(&ConfigSet::from(config.clone()));
+
+                    if !config.links.is_empty() {
+                        config_set = config_set
+                            .merge_right(&self.read_links(ConfigSet::from(config)).await?);
+                    }
+                }
+                LinkType::GraphQL => {
+                    let service_document =
+                        Config::from_source(Source::detect(&source.path).unwrap(), &content)?
```
Comment (tusharmath) [inline src/config/reader.rs:99]: May be we can drop the `GraphQL` type also for now.

File: src/config/reader.rs [hunk 5]
```diff
@@ -107,6 +181,9 @@ impl ConfigReader {
         // Extend it with protobuf definitions for GRPC
         let config_set = self.ext_grpc(config_set).await?;
 
+        // Extend it with the links
+        let config_set = self.read_links(config_set).await?;
```
Comment (tusharmath) [inline src/config/reader.rs:185]:
```suggestion
        let config_set = self.ext_links(config_set).await?;
```

File: src/config/reader.rs [hunk 6]
```diff
@@ -62,6 +65,77 @@ impl ConfigReader {
...
+
+                    config_set
+                        .extensions
+                        .service_document_from_links
+                        .push(ServiceDocumentWithId {
+                            id: config_link.id.to_owned(),
+                            service_document,
+                        });
+                }
+                LinkType::Protobuf => {
```
Comment (tusharmath) [inline src/config/reader.rs:110]: With this feature, we should drop the `protoPath` that's available on the GRPC operator.

File: src/config/from_document.rs [hunk 7]
```diff
@@ -75,15 +78,66 @@ fn process_schema_directives<T: DirectiveCodec<T> + Default>(
...
+    let mut res = Valid::succeed(Vec::new());
+
+    for directive in directives {
+        res = res.zip(directive).map(|(mut vec, item)| {
+            vec.push(item);
+            vec
+        });
+    }
+
+    res
```
Comment (tusharmath) [inline src/config/from_document.rs:106]: Can use Valid::from_iter

File: src/config/from_document.rs [hunk 8]
```diff
@@ -75,15 +78,66 @@ fn process_schema_directives<T: DirectiveCodec<T> + Default>(
...
+        for link in &links {
+            if let Some(id) = &link.id {
+                if ids.contains(id) {
+                    return Valid::fail(format!("Duplicated id: {}", id))
+                        .trace(config::Link::directive_name().as_str());
+                }
+                ids.push(id.to_string());
+            }
+        }
+        Valid::succeed(links)
```
Comment (tusharmath) [inline src/config/from_document.rs:137]: This logic will bail out on the first duplicate found. 
IMPORTANT - Write a test case to check if multiple duplicates are reported.

File: src/config/config_set.rs [hunk 9]
```diff
@@ -11,6 +12,34 @@ pub struct ConfigSet {
...
+            service_document: ServiceDocument { definitions: vec![] },
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct RawContent {
+    pub id: Option<String>,
+    pub content: async_graphql_value::ConstValue,
+}
```
Comment (tusharmath) [inline src/config/config_set.rs:41]: I think we can create a polymorphic container 

```rust
pub struct Asset<A> {
    id: Option<string>,
    content: A
}
```

File: src/config/from_document.rs (hunk 8 above)
Comment (ologbonowiwi) [reply src/config/from_document.rs:137]: [tests/graphql/errors/test-link.graphql](https://github.com/tailcallhq/tailcall/pull/1094/files#diff-11d605567b472743fd4e0d46c87318f6f1ca28a6eff75d0307f26164ce8510d7) covers this, I'll change its name to make clearer

File: src/config/config_set.rs (hunk 9 above)
Comment (ologbonowiwi) [reply src/config/config_set.rs:41]: we don't need it anymore as we won't support `Data` (this was only used by `Data` type

File: src/config/from_document.rs (hunk 8 above)
Comment (ologbonowiwi) [reply src/config/from_document.rs:128]: test renamed

File: src/config/config_set.rs (hunk 9 above)
Comment (ologbonowiwi) [reply src/config/config_set.rs]: now I got it

Comment (ologbonowiwi) [reply src/config/config_set.rs]: I've used the `Content` name, but let me know if you prefer `Asset`; also, great suggestion, thank you :smile:

File: tests/graphql/test-add-link-to-empty-config.graphql (hunk 3 above)
Comment (ologbonowiwi) [reply tests/graphql/test-add-link-to-empty-config.graphql]: done

File: examples/grpc.graphql [hunk 10]
```diff
@@ -1,20 +1,21 @@
...
   newsById(news: NewsInput!): News!
-    @grpc(service: "news.NewsService", method: "GetNews", body: "{{args.news}}", protoPath: "src/grpc/tests/news.proto")
+    @grpc(service: "news.NewsService", method: "GetNews", body: "{{args.news}}", protoId: "news")
   newsByIdBatch(news: NewsInput!): News!
     @grpc(
       service: "news.NewsService"
       method: "GetMultipleNews"
       body: "{{args.news}}"
-      protoPath: "src/grpc/tests/news.proto"
+      protoId: "news"
```
Comment (tusharmath) [inline examples/grpc.graphql:18]: I feel we dont need protoId at all anymore. Since the buf files are loaded via link already we can just keep it simple by having a service, method and body. Infact we could also do just method. For eg:

**before**
```graphql
@grpc(service: "news.NewsService", method: "GetMultipleNews", body: "{{args.news}}", protoId: "news")
```

**After**
```graphql
@grpc(method: "news.NewsService.GetMultipleNews", body: "{{args.news}}")
```

IMO, this is a much nicer and a slicker user experience, makes me want to use the feature :)

Comment (tusharmath) [reply examples/grpc.graphql:18]: Feel free to create a separate issue if you think it's going out of scope :)

Comment (ologbonowiwi) [reply examples/grpc.graphql:18]: Sure! #1101 created

File: src/config/config_set.rs [hunk 11]
```diff
@@ -1,41 +1,62 @@
...
 use crate::config::Config;
 
 /// A wrapper on top of Config that contains all the resolved extensions.
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug, Default, Setters)]
 pub struct ConfigSet {
     pub config: Config,
+    // Contains the file path of the file from which the config was read
+    // In case multiple files are read, the last file is stored
+    pub path: Option<String>,
```
Comment (tusharmath) [inline src/config/config_set.rs:14]: Why do we need path here?

File: src/config/config_set.rs [hunk 12]
```diff
@@ -1,41 +1,62 @@
...
+
 /// Extensions are meta-information required before we can generate the blueprint.
 /// Typically, this information cannot be inferred without performing an IO operation, i.e.,
 /// reading a file, making an HTTP call, etc.
 #[derive(Clone, Debug, Default)]
 pub struct Extensions {
-    pub grpc_file_descriptor: FileDescriptorSet,
-
-    /// Contains the contents of the JS file
-    pub script: Option<String>,
```
Comment (tusharmath) [inline src/config/config_set.rs:22]: Why is the script deleted?

File: src/config/from_document.rs [hunk 13]
```diff
@@ -75,15 +78,57 @@ fn process_schema_directives<T: DirectiveCodec<T> + Default>(
     res
 }
 
+fn process_schema_multiple_directives<T: DirectiveCodec<T> + Default>(
```
Comment (tusharmath) [inline src/config/from_document.rs:81]: We can avoid duplication right with `process_schema_directives`. I think having one of the two functions is better.

File: src/config/config_set.rs [hunk 14]
```diff
@@ -1,41 +1,62 @@
...
+    /// Contains the file descriptor sets resolved from the links
+    pub file_descriptor_from_links: Vec<Content<FileDescriptorSet>>,
```
Comment (tusharmath) [inline src/config/config_set.rs:37]:
```suggestion
    pub grpc_file_descriptors: Vec<Content<FileDescriptorSet>>,
```
Simpler name perhaps.

File: src/config/link.rs [hunk 15]
```diff
@@ -0,0 +1,41 @@
+use std::fmt::Display;
+
+use serde::{Deserialize, Serialize};
+
+use super::super::is_default;
+
+#[derive(Default, Serialize, Deserialize, PartialEq, Eq, Debug, Clone, schemars::JsonSchema)]
+pub enum LinkType {
+    #[default]
```
Comment (tusharmath) [inline src/config/link.rs:9]: do we need a default?

File: generated/.tailcallrc.schema.json [hunk 16]
```diff
@@ -1192,6 +1192,38 @@
       },
       "type": "object"
     },
+    "Link": {
```
Comment (tusharmath) [inline generated/.tailcallrc.schema.json:1195]: Need to add `Link` to the whitelist of type that we want to add in tailcallrc.graphql. The code is in autogen.

File: src/config/link.rs (hunk 15 above)
Comment (tusharmath) [reply src/config/link.rs:9]: May be it's a good idea, never mind. It's a great convinence feature :)

File: src/config/link.rs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    Config,
+    Protobuf,
+}
+
+impl Display for LinkType {
```
Comment (tusharmath) [inline src/config/link.rs:14]: This we can derive, can't we?

File: src/config/from_document.rs (hunk 13 above)
Comment (ologbonowiwi) [reply src/config/from_document.rs:81]: essentially, these two methods are doing different things. 

`process_schema_directives` handles one specific directive and uses the last occurrence of it. `process_schema_multiple_directives`, by the other side, returns all directives with this occurrence.
I can use only one of'em and add a new boolean value to check if it should behave as `process_schema_multiple_directives` or `process_schema_directives`, but IMO this should be avoided.

plase let me know if you have something else in mind

File: src/config/link.rs (hunk 17 above)
Comment (ologbonowiwi) [reply src/config/link.rs:14]: it does not work for enum, optionally we can install an external crate like [enum_display_derive](https://docs.rs/enum-display-derive/latest/enum_display_derive/) or [strum](https://docs.rs/strum/0.26.1/strum/)

File: src/config/reader.rs [hunk 18]
```diff
@@ -62,6 +62,74 @@ impl ConfigReader {
...
+            let source = if let Ok(data) = self.read_file(&config_link.src).await {
+                data
+            } else {
+                let path = path.clone().unwrap_or_default();
+                let path = PathBuf::from(&path)
+                    .parent()
+                    .unwrap_or(Path::new(""))
+                    .join(&config_link.src);
+
+                self.read_file(path.to_string_lossy()).await?
```
Comment (tusharmath) [inline src/config/reader.rs:92]: Why are we reading the file twice here?

File: src/config/reader.rs [hunk 19]
```diff
@@ -62,6 +62,74 @@ impl ConfigReader {
...
+            };
+
+            let content = source.content;
+
+            match config_link.type_of {
+                LinkType::Config => {
+                    let config =
+                        Config::from_source(Source::detect(&source.path).unwrap(), &content)?;
```
Comment (tusharmath) [inline src/config/reader.rs:100]: can't use unwrap.

File: src/config/reader.rs (hunk 18 above)
Comment (ologbonowiwi) [reply src/config/reader.rs:92]: if the first read (using absolute path) fails, we go for relative path

File: src/config/reader.rs (hunk 19 above)
Comment (ologbonowiwi) [reply src/config/reader.rs]: fixed

File: tests/snapshots/execution_spec__test-empty-link.md_errors.snap [hunk 20]
```diff
@@ -0,0 +1,11 @@
+---
+source: tests/execution_spec.rs
+expression: errors
+---
+[
+  {
+    "message": "Link src cannot be empty",
+    "trace": [],
```
Comment (tusharmath) [inline tests/snapshots/execution_spec__test-empty-link.md_errors.snap:8]: trace should contain `["schema", "@link"]`

File: tests/execution/test-duplicated-link.md [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+###### sdl error
+
+#### server:
+
+```graphql
+schema
+  @link(type: Config, src: "examples/jsonplaceholder.graphql", id: "placeholder")
+  @link(type: Config, src: "examples/jsonplaceholder.graphql", id: "placeholder1")
+  @link(type: Config, src: "examples/jsonplaceholder.graphql", id: "placeholder1") {
+  query: Query
```
Comment (tusharmath) [inline tests/execution/test-duplicated-link.md:12]: Let's add some tests with multiple duplicates here.

File: src/blueprint/links.rs [hunk 22]
```diff
@@ -0,0 +1,37 @@
+use crate::config::Link;
+use crate::directive::DirectiveCodec;
+use crate::valid::{Valid, ValidationError, Validator};
+
+// Wrap the Vec<Link> is necessary to implement TryFrom
+#[allow(dead_code)]
+pub struct Links(Vec<Link>);
```
Comment (ologbonowiwi) [inline src/blueprint/links.rs:7]: Creating this was necessary as `TryFrom` can't be directly implemented for `Vec` as both the trait and the struct are external types.

File: src/blueprint/links.rs [hunk 23]
```diff
@@ -0,0 +1,37 @@
+use crate::config::Link;
+use crate::directive::DirectiveCodec;
+use crate::valid::{Valid, ValidationError, Validator};
+
+// Wrap the Vec<Link> is necessary to implement TryFrom
+#[allow(dead_code)]
```
Comment (ologbonowiwi) [inline src/blueprint/links.rs:6]: This is necessary as `(Vec<Link>)` is not used anywhere in the code.

Optionally, we can remove it and change `Links` to be an empty struct – `pub struct Links` – or a unit type – – `pub struct Links(())` –.

File: src/blueprint/from_config.rs [hunk 24]
```diff
@@ -30,10 +30,15 @@ pub fn config_blueprint<'a>() -> TryFold<'a, ConfigSet, Blueprint, String> {
             .map(|upstream| blueprint.upstream(upstream))
     });
 
+    let links = TryFoldConfig::<Blueprint>::new(|config_set, blueprint| {
+        Valid::from(Links::try_from(config_set.links.clone())).map_to(blueprint)
+    });
```
Comment (ologbonowiwi) [inline src/blueprint/from_config.rs:35]: Used `Links::try_from` here to keep the standard, but `map_to(blueprint)` as we don't links property on blueprint.

**Review by tusharmath — APPROVED** [review]

File: src/blueprint/from_config.rs (hunk 24 above)
Comment (tusharmath) [reply src/blueprint/from_config.rs:35]: I fixed it.

File: src/config/reader.rs (hunk 18 above)
Comment (tusharmath) [reply src/config/reader.rs:99]: We shouldn't do that. This should ideally work like how node imports modules. Essentially, if the absolute path exists, use that or if it's a relative path, create an equivalent absolute URL based on the current file path.
